### DISCUSSION TITLE
Support digest updates for go mod pseudo versions for latest untagged commit

### DISCUSSION THREAD

**Discussion by omercnet** [renovatebot/renovate#24841] [Suggest an Idea]
### What would you like Renovate to be able to do?

When setting `digest: true` for go modules, renovate will run `go get -u` but per `go get` behavior it will not not look for digest release if the current version is tagged.

To get the latest digest, `go mod` must target a git branch from which to get the latest digest.

see https://github.com/golang/go/wiki/Modules#how-to-upgrade-and-downgrade-dependencies for details

example
```zsh
❯ go get -u github.com/omercnet/go-untagged     # gets latest *tag*
go: added github.com/omercnet/go-untagged v0.0.2

❯ go get -u github.com/omercnet/go-untagged@main # gets latest *commit*
go: upgraded github.com/omercnet/go-untagged v0.0.2 => v0.0.3-0.20221125140523-6596438694dc

❯ go get -u github.com/omercnet/go-untagged     # does nothing once already on latest *commit*

❯ go get -u github.com/omercnet/go-untagged@latest # reverts back to latest *tag*
go: downgraded github.com/omercnet/go-untagged v0.0.3-0.20221125140523-6596438694dc => v0.0.2

❯ go get -u github.com/omercnet/go-untagged       # does nothing as already on latest tag

❯ go get -u github.com/omercnet/go-untagged@main  # gets latest *commit* 
go: upgraded github.com/omercnet/go-untagged v0.0.2 => v0.0.3-0.20221125140523-6596438694dc
```

### If you have any ideas on how this should be implemented, please tell us here.

when `digest: true` is specified on a gomod dependency, renovate should execute `go get <dep>@<branch>` where the default branch is `main`, and introduce a new configuration setting to allow customizing the branch from which to fetch the latest commit

### Is this a feature you are interested in implementing yourself?

Yes

**Comment by rarkins** [comment]
Where is this code currently?

**Comment by omercnet** [comment]
Which code?

**Comment by rarkins** [comment]
The "renovate will run" part. When I search the code base there's no hit for the string `mod -u`

**Comment by omercnet** [comment]
it's `go get -u`, sorry
here's the code 
https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/gomod/artifacts.ts#L275

**Comment by rarkins** [comment]
Do we already know what the default branch is? And is this meant to work across all platforms, not just github?

**Comment by omercnet** [comment]
As far as I understand, it works with all git backends that go supports

We don't know the default branch, needs to be configurable

**Comment by rarkins** [comment]
So will you need an initial command to learn the default branch?

**Comment by omercnet** [comment]
Don't think it's possible to learn
Have a sane default to main/master and expose a new configuration key to specify it

**Comment by rarkins** [comment]
It may be possible, but it's not elegant: https://davidwalsh.name/get-default-branch-name

**Comment by viceice** [comment]
why do we need the default branch? can't we just use `HEAD` ?

**Comment by omercnet** [comment]
that's a good point @viceice we can just use `@HEAD` !

**Comment by G-Rath** [comment]
Just wondering if there was any ever work done on this? it sounds exactly like what we're wanting over in `osv-scanner`, to have Renovate maintain a PR updating `osv-scalibr` to the latest _commit_ on its default branch to help ensure nothing has been broken, and to be occasionally merged to actually update the version we're using.

Currently this looks to working so long as we're using the latest tagged version in our pseudo version but when a new version is tagged Renovate starts targeting that as the latest which sounds like what is being described here
